File: src/backtesting/validator.py
"""Statistical validation of backtest results."""

from __future__ import annotations

import math
import random
from decimal import Decimal
from typing import Any

from pydantic import BaseModel

from src.core.logging import get_logger

log = get_logger(__name__)

_ZERO = Decimal("0")


class ValidationResult(BaseModel):
    """Result of statistical validation."""

    sharpe_ci_95: tuple[float, float]
    p_value: float
    is_significant: bool
    overfitting_warning: str | None = None

    model_config = {"frozen": True}


class StatisticalValidator:
    """Validate that backtest results are statistically significant."""

    def __init__(
        self,
        n_bootstrap: int = 1000,
        n_permutations: int = 1000,
        significance_level: float = 0.05,
        rng_seed: int | None = None,
    ) -> None:
        self._n_bootstrap = n_bootstrap
        self._n_permutations = n_permutations
        self._significance_level = significance_level
        self._rng = random.Random(rng_seed)

    def validate(
        self,
        trades: list[dict[str, Any]],
        equity_curve: list[Decimal],
    ) -> ValidationResult:
        """Run all validation checks.

        Args:
            trades: List of trade dicts with 'pnl' key.
            equity_curve: Portfolio equity over time.

        Returns:
            ValidationResult with CI, p-value, and warnings.
        """
        returns = self._equity_to_returns(equity_curve)
        sharpe_ci = self._bootstrap_sharpe_ci(returns)
        p_value = self._permutation_test(trades)
        is_significant = p_value < self._significance_level

        # Overfitting check: split equity curve 70/30 and compare Sharpe
        overfitting_warning: str | None = None
        if len(equity_curve) >= 20:
            split_idx = int(len(equity_curve) * 0.7)
            is_returns = self._equity_to_returns(equity_curve[:split_idx])
            oos_returns = self._equity_to_returns(equity_curve[split_idx:])
            if len(is_returns) >= 2 and len(oos_returns) >= 2:
                is_sharpe = self._sharpe(is_returns)
                oos_sharpe = self._sharpe(oos_returns)
                overfitting_warning = self.check_overfitting(is_sharpe, oos_sharpe)

        log.info(
            "validation_complete",
            sharpe_ci_low=f"{sharpe_ci[0]:.4f}",
            sharpe_ci_high=f"{sharpe_ci[1]:.4f}",
            p_value=f"{p_value:.4f}",
            significant=is_significant,
            overfitting=overfitting_warning or "none",
        )

        return ValidationResult(
            sharpe_ci_95=sharpe_ci,
            p_value=p_value,
            is_significant=is_significant,
            overfitting_warning=overfitting_warning,
        )

    def check_overfitting(
        self,
        in_sample_sharpe: float,
        out_of_sample_sharpe: float,
    ) -> str | None:
        """Check for overfitting by comparing IS vs OOS Sharpe ratios.

        Args:
            in_sample_sharpe: Sharpe ratio on training data.
            out_of_sample_sharpe: Sharpe ratio on holdout data.

        Returns:
            Warning string if overfitting detected, else None.
        """
        if out_of_sample_sharpe <= 0:
            warning = (
                f"Out-of-sample Sharpe ({out_of_sample_sharpe:.2f}) is non-positive. "
                f"In-sample Sharpe was {in_sample_sharpe:.2f}. Likely overfitting."
            )
            log.warning("overfitting_detected", warning=warning)
            return warning

        ratio = in_sample_sharpe / out_of_sample_sharpe
        if ratio > 2.0:
            warning = (
                f"IS/OOS Sharpe ratio is {ratio:.2f}x "
                f"(IS={in_sample_sharpe:.2f}, OOS={out_of_sample_sharpe:.2f}). "
                f"Possible overfitting."
            )
            log.warning("overfitting_detected", warning=warning)
            return warning

        return None

    def _bootstrap_sharpe_ci(
        self,
        returns: list[float],
    ) -> tuple[float, float]:
        """Bootstrap 95% confidence interval on Sharpe ratio.

        Args:
            returns: Period returns.

        Returns:
            (lower, upper) bounds of 95% CI.
        """
        if len(returns) < 2:
            return (0.0, 0.0)

        n = len(returns)
        bootstrapped_sharpes: list[float] = []

        for _ in range(self._n_bootstrap):
            sample = [self._rng.choice(returns) for _ in range(n)]
            sharpe = self._sharpe(sample)
            bootstrapped_sharpes.append(sharpe)

        bootstrapped_sharpes.sort()
        lower_idx = int(self._n_bootstrap * 0.025)
        upper_idx = int(self._n_bootstrap * 0.975)
        return (bootstrapped_sharpes[lower_idx], bootstrapped_sharpes[upper_idx])

    def _permutation_test(
        self,
        trades: list[dict[str, Any]],
    ) -> float:
        """Monte Carlo permutation test on trade PnLs.

        Shuffles trade PnLs and computes how often the shuffled total
        exceeds the observed total. Returns a p-value.

        Args:
            trades: List of trade dicts with 'pnl' key.

        Returns:
            p-value (proportion of permutations >= observed).
        """
        if not trades:
            return 1.0

        pnls = [float(t["pnl"]) for t in trades]
        observed_total = sum(pnls)
        count_gte = 0

        for _ in range(self._n_permutations):
            shuffled = list(pnls)
            # Randomly flip signs to create null distribution
            shuffled = [p * self._rng.choice([1, -1]) for p in shuffled]
            if sum(shuffled) >= observed_total:
                count_gte += 1

        return count_gte / self._n_permutations

    @staticmethod
    def _equity_to_returns(equity_curve: list[Decimal]) -> list[float]:
        """Convert equity curve to period returns."""
        if len(equity_curve) < 2:
            return []
        returns: list[float] = []
        for i in range(1, len(equity_curve)):
            prev = equity_curve[i - 1]
            if prev == _ZERO:
                returns.append(0.0)
            else:
                returns.append(float((equity_curve[i] - prev) / prev))
        return returns

    @staticmethod
    def _sharpe(returns: list[float]) -> float:
        """Compute annualized Sharpe ratio from returns."""
        if len(returns) < 2:
            return 0.0
        mean_r = sum(returns) / len(returns)
        variance = sum((r - mean_r) ** 2 for r in returns) / len(returns)
        std_r = math.sqrt(variance)
        if std_r == 0.0:
            return 0.0
        return mean_r / std_r * math.sqrt(252)

File: rust-backtest/crates/mvhe-engine/src/validator.rs
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use rayon::prelude::*;
use serde::Serialize;

use crate::metrics::sharpe_from_returns;

/// Result of statistical validation.
#[derive(Debug, Clone, Serialize)]
pub struct ValidationResult {
    pub sharpe_ci_95: (f64, f64),
    pub p_value: f64,
    pub is_significant: bool,
    pub overfitting_warning: Option<String>,
}

/// Statistical validator for backtest results.
///
/// Validates that backtest performance is statistically significant using:
/// - Bootstrap 95% CI on Sharpe ratio (parallelized with Rayon)
/// - Permutation test on trade PnLs (parallelized with Rayon)
/// - IS/OOS overfitting check (70/30 split)
pub struct StatisticalValidator {
    n_bootstrap: usize,
    n_permutations: usize,
    significance_level: f64,
    seed: u64,
}

impl StatisticalValidator {
    pub fn new(
        n_bootstrap: usize,
        n_permutations: usize,
        significance_level: f64,
        seed: u64,
    ) -> Self {
        Self {
            n_bootstrap,
            n_permutations,
            significance_level,
            seed,
        }
    }

    /// Run all validation checks.
    pub fn validate(
        &self,
        equity_curve: &[f64],
        trade_pnls: &[f64],
    ) -> ValidationResult {
        let returns = Self::equity_to_returns(equity_curve);

        let sharpe_ci = self.bootstrap_sharpe_ci(&returns);
        let p_value = self.permutation_test(trade_pnls);
        let is_significant = p_value < self.significance_level;

        // Overfitting check
        let overfitting_warning = if equity_curve.len() >= 20 {
            let split_idx = (equity_curve.len() as f64 * 0.7) as usize;
            let is_returns = Self::equity_to_returns(&equity_curve[..split_idx]);
            let oos_returns = Self::equity_to_returns(&equity_curve[split_idx..]);
            if is_returns.len() >= 2 && oos_returns.len() >= 2 {
                let is_sharpe = sharpe_from_returns(&is_returns);
                let oos_sharpe = sharpe_from_returns(&oos_returns);
                Self::check_overfitting(is_sharpe, oos_sharpe)
            } else {
                None
            }
        } else {
            None
        };

        ValidationResult {
            sharpe_ci_95: sharpe_ci,
            p_value,
            is_significant,
            overfitting_warning,
        }
    }

    /// Bootstrap 95% CI on Sharpe ratio using Rayon parallelism.
    fn bootstrap_sharpe_ci(&self, returns: &[f64]) -> (f64, f64) {
        if returns.len() < 2 {
            return (0.0, 0.0);
        }

        let n = returns.len();
        let seed = self.seed;
        let n_bootstrap = self.n_bootstrap;

        // Each bootstrap iteration uses an independent RNG seeded deterministically
        let mut bootstrapped_sharpes: Vec<f64> = (0..n_bootstrap)
            .into_par_iter()
            .map(|i| {
                let mut rng = StdRng::seed_from_u64(seed.wrapping_add(i as u64));
                let sample: Vec<f64> = (0..n).map(|_| returns[rng.gen_range(0..n)]).collect();
                sharpe_from_returns(&sample)
            })
            .collect();

        bootstrapped_sharpes.sort_unstable_by(|a, b| a.partial_cmp(b).unwrap());

        let lower_idx = (n_bootstrap as f64 * 0.025) as usize;
        let upper_idx = (n_bootstrap as f64 * 0.975) as usize;

        (
            bootstrapped_sharpes[lower_idx],
            bootstrapped_sharpes[upper_idx.min(bootstrapped_sharpes.len() - 1)],
        )
    }

    /// Monte Carlo permutation test: randomly flip PnL signs.
    fn permutation_test(&self, pnls: &[f64]) -> f64 {
        if pnls.is_empty() {
            return 1.0;
        }

        let observed_total: f64 = pnls.iter().sum();
        let seed = self.seed;
        let n_permutations = self.n_permutations;

        let count_gte: usize = (0..n_permutations)
            .into_par_iter()
            .filter(|&i| {
                let mut rng = StdRng::seed_from_u64(seed.wrapping_add(10000 + i as u64));
                let shuffled_total: f64 = pnls
                    .iter()
                    .map(|&p| {
                        if rng.gen_bool(0.5) {
                            p
                        } else {
                            -p
                        }
                    })
                    .sum();
                shuffled_total >= observed_total
            })
            .count();

        count_gte as f64 / n_permutations as f64
    }

    fn check_overfitting(is_sharpe: f64, oos_sharpe: f64) -> Option<String> {
        if oos_sharpe <= 0.0 {
            return Some(format!(
                "Out-of-sample Sharpe ({:.2}) is non-positive. In-sample was {:.2}. Likely overfitting.",
                oos_sharpe, is_sharpe
            ));
        }
        let ratio = is_sharpe / oos_sharpe;
        if ratio > 2.0 {
            return Some(format!(
                "IS/OOS Sharpe ratio is {:.2}x (IS={:.2}, OOS={:.2}). Possible overfitting.",
                ratio, is_sharpe, oos_sharpe
            ));
        }
        None
    }

    fn equity_to_returns(equity_curve: &[f64]) -> Vec<f64> {
        if equity_curve.len() < 2 {
            return Vec::new();
        }
        equity_curve
            .windows(2)
            .map(|w| {
                if w[0] == 0.0 {
                    0.0
                } else {
                    (w[1] - w[0]) / w[0]
                }
            })
            .collect()
    }
}

impl Default for StatisticalValidator {
    fn default() -> Self {
        Self::new(1000, 1000, 0.05, 42)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_bootstrap_ci_empty() {
        let validator = StatisticalValidator::default();
        let (lo, hi) = validator.bootstrap_sharpe_ci(&[]);
        assert_eq!(lo, 0.0);
        assert_eq!(hi, 0.0);
    }

    #[test]
    fn test_permutation_test_zero_pnl() {
        let validator = StatisticalValidator::default();
        let pnls = vec![0.0, 0.0, 0.0];
        let p = validator.permutation_test(&pnls);
        // With all zeros, every permutation matches -> p_value = 1.0
        assert!((p - 1.0).abs() < 0.1);
    }

    #[test]
    fn test_permutation_test_significant() {
        let validator = StatisticalValidator::new(100, 1000, 0.05, 42);
        // Very profitable strategy
        let pnls = vec![10.0; 50];
        let p = validator.permutation_test(&pnls);
        assert!(p < 0.05, "Expected significant p-value, got {}", p);
    }

    #[test]
    fn test_check_overfitting_none() {
        let result = StatisticalValidator::check_overfitting(2.0, 1.5);
        assert!(result.is_none());
    }

    #[test]
    fn test_check_overfitting_detected() {
        let result = StatisticalValidator::check_overfitting(4.0, 1.0);
        assert!(result.is_some());
    }

    #[test]
    fn test_validate_full() {
        let validator = StatisticalValidator::new(100, 100, 0.05, 42);
        let equity: Vec<f64> = (0..100).map(|i| 10000.0 + i as f64 * 10.0).collect();
        let pnls: Vec<f64> = (0..50).map(|_| 20.0).collect();

        let result = validator.validate(&equity, &pnls);
        assert!(result.sharpe_ci_95.0 <= result.sharpe_ci_95.1);
    }
}
